slang.h

```diff
@@ -450,6 +450,178 @@ convention for interface methods.
 #include <stddef.h>
 #endif // ! SLANG_NO_STDDEF
 
+namespace slang
+{
+
+/* Slang provides a mechanism for value types (such as structures) to be able to provide forward and backward
+ABI compatibility through. 
+
+The type is made up of several parts.
+0) The kind primary or extension
+1) The category the type is in
+2) An id for the type in the category
+3) The major semantic version
+4) The minor semantic version
+
+Within the 'semantic versioning' a structure that has only a larger minor semantic version can be used (with some
+caveats) as is when being passed to a previous version of Slang. It may contain more fields.
+*/
+
+enum class StructTagKind : uint8_t
+{
+    Primary,                ///< A primary struct (contains optional extensions)
+    Extension,              ///< An extension struct
+};
+
+#define SLANG_STRUCT_TAG_ENUM(x) x,
+
+#define SLANG_STRUCT_TAG_CATEGORIES(x) \
+    x(Core) \
+    x(Slang) \
+    x(Gfx)
+
+
+enum class StructTagCategory : uint8_t
+{
+    SLANG_STRUCT_TAG_CATEGORIES(SLANG_STRUCT_TAG_ENUM)
+    CountOf,
+};
+
+/* AbiStructType is laid out as follows
+
+     |Primary | Category     | Type      | Major version | Minor Version
+-----|--------|--------------|-----------|---------------|----------------
+Bits | 31     | 30-24        | 23-16     |     15-8      |     0-7
+Type |        | AbiCategory  | Cat Spec  |               |
+
+Type will be specified as something specific for the category. A typical implementation will have an enum that lists
+types for that category. 
+
+TODO(JS): This layout may need to be altered - too may bits are perhaps used by major/minor version for example. This should
+be fine for the immediate future though.
```

Comment:
One note is that the `Primary`-ness of a structure may not need to be encoded as its own bit, since it would be implied by the `Type`. The bit is thus redundant for any recognized `Category`+`Type` pair, and for any case where we don't recognize the category and/or type, just knowing whether something is primary or not is only of limited value (we could specialize a diagnostic message based on that bit, but I'm not clear it is worth distinguishing from other "I don't recognize this type" cases).

Reply:
(I was about to put forward a bunch more ideas on how we could allocate the bits for all of this, but a PR review is probably not the best place to bikeshed all the design stuff)